rest-service/manager_rest/workflow_executor.py

```diff
@@ -252,3 +252,54 @@ def cancel_execution(execution_id):
         }
     }
     _send_mgmtworker_task(message, routing_key='service')
+
+
+def _get_install_plugin_message(plugin, target_names=None):
+    sm = get_storage_manager()
+    managers = sm.list(models.Manager)
+    message = {
+        'service_task': {
+            'task_name': 'install-plugin',
+            'kwargs': {
+                'plugin': plugin.to_dict(),
+                'rest_host': [manager.private_ip for manager in managers],
+                'rest_token': current_user.get_auth_token(),
+                'tenant': _get_tenant_dict(),
+            }
+        }
+    }
+    if target_names:
+        message['service_task']['kwargs']['target'] = target_names
+    return message
+
+
+def install_plugin_on_managers(plugin, target_names):
+    _send_mgmtworker_task(
+        _get_install_plugin_message(plugin, target_names=None),
+        routing_key='service')
+
+
+def install_plugin_on_agents(plugin, target_agents):
+    # all the agents are supposed to be on the same tenant, so we only need
+    # one amqp client, but let's check that. There should never be a way
+    # to get agents from multiple tenants into this function.
```

Comment:
What about with deployments with visibility: shared?

Reply:
good point. Well, disallowed for now. But I've pushed a commit that allows installing plugins on those, even if you need N requests to install on agents from N tenants.

Reply:
Sounds reasonable on this pass.